traits/trait_types.pyi

```diff
@@ -494,24 +498,70 @@ class BaseClass(_BaseClass[_Type[_Any]]):
     ...
 
 
-class _BaseInstance(_BaseClass[_T]):
+class BaseInstance(_TraitType[_S, _S]):
+
+    # simplified signatures
 
-    # simplified signature
+    @overload
     def __init__(
-        self,
-        klass: _T,
+        self: BaseInstance[Optional[_T]],
+        klass: _Type[_T],
         *args: _Any,
+        allow_none: Literal[True] = ...,
         **metadata: _Any,
     ) -> None:
         ...
 
+    @overload
+    def __init__(
+        self: BaseInstance[_T],
+        klass: _Type[_T],
+        *args: _Any,
+        allow_none: Literal[False] = ...,
+        **metadata: _Any,
+    ) -> None:
+        ...
 
-class BaseInstance(_BaseInstance[_Any]):
-    ...
+    @overload
+    def __init__(
+        self: BaseInstance[_Any],
+        klass: str,
+        *args: _Any,
+        **metadata: _Any,
+    ) -> None:
+        ...
 
 
-class Instance(_BaseInstance[_Any]):
-    ...
+class Instance(BaseInstance[_S]):
+
+    @overload
```

Comment:
I couldn't find a way not to have to repeat these overloads - simply doing `class Instance(BaseInstance): ...` led to anything being accepted for the trait values. Hints welcome.